src/platform/nrfconnect/wifi/WiFiManager.cpp

```diff
@@ -272,35 +276,91 @@ CHIP_ERROR WiFiManager::GetNetworkStatistics(NetworkStatistics & stats) const
+            // Fallback to the WIFI_SECURITY_TYPE_PSK if the security is unknown
+            Instance().mWiFiParams.mParams.security =
+                scanResult->security <= WIFI_SECURITY_TYPE_MAX ? scanResult->security : WIFI_SECURITY_TYPE_PSK;
+            Instance().mWiFiParams.mParams.psk_length = Instance().mCurrentCredentials.size();
+
+            // If the security is none, WiFi driver expects the psk to be nullptr
+            if (Instance().mWiFiParams.mParams.security == WIFI_SECURITY_TYPE_NONE)
+            {
+                Instance().mWiFiParams.mParams.psk = nullptr;
+            }
+            else
+            {
+                Instance().mWiFiParams.mParams.psk = const_cast<uint8_t *>(Instance().mCurrentCredentials.data());
+            }
+
+            Instance().mWiFiParams.mParams.timeout = Instance().mHandling.mConnectionTimeoutMs.count();
+            Instance().mWiFiParams.mParams.channel = scanResult->channel;
+            Instance().mWiFiParams.mRssi           = scanResult->rssi;
+        }
+    }
+
+    if (Instance().mScanResultCallback && !Instance().mInternalScan)
+    {
+        NetworkCommissioning::WiFiScanResponse response = ToScanResponse(scanResult);
+        Instance().mScanResultCallback(scanResult != nullptr ? &response : nullptr);
+    }
 }
 
-void WiFiManager::ScanDoneClbk(NetEventCallback * cb)
+void WiFiManager::ScanDoneHandler(uint8_t * data)
 {
-    VerifyOrReturn(Instance().mScanDoneCallback != nullptr);
-    const WiFiStatus * status       = static_cast<const WiFiStatus *>(cb->info);
+    const WiFiStatus * status       = reinterpret_cast<const WiFiStatus *>(data);
     WiFiRequestStatus requestStatus = static_cast<WiFiRequestStatus>(status->status);
 
+    if (Instance().mScanDoneCallback && !Instance().mInternalScan)
+    {
+        Instance().mScanDoneCallback(requestStatus);
+    }
+
     if (requestStatus == WiFiRequestStatus::FAIL)
     {
         ChipLogDetail(DeviceLayer, "Scan request failed (%d)", status->status);
     }
     else
     {
         ChipLogDetail(DeviceLayer, "Scan request done (%d)", status->status);
-    }
 
-    Instance().mScanDoneCallback(requestStatus);
+        // Internal scan is supposed to be followed by connection request
+        if (Instance().mInternalScan)
+        {
+            Instance().mWiFiState = WIFI_STATE_ASSOCIATING;
+            NetInterfacePtr iface = InetUtils::GetInterface();
+            VerifyOrReturn(nullptr != iface, CHIP_ERROR_INTERNAL);
+
+            if (net_mgmt(NET_REQUEST_WIFI_CONNECT, iface, &(Instance().mWiFiParams.mParams), sizeof(wifi_connect_req_params)))
+            {
+                ChipLogError(DeviceLayer, "Connection request failed");
+                if (Instance().mHandling.mOnConnectionFailed)
+                {
+                    Instance().mHandling.mOnConnectionFailed();
+                }
+            }
+            ChipLogError(DeviceLayer, "Connection to %s requested", Instance().mWiFiParams.mParams.ssid);
```

Comment:
This will be printed even in the case of error. Missing `return` above?

Reply:
Yes, thanks.